subsys/net/lib/dns/llmnr_responder.c

```diff
@@ -0,0 +1,610 @@
+{
+	u16_t flags;
+
+	/* See RFC 1035, ch 4.1.1 and RFC 4795 ch 2.1.1 for header details */
+
+	flags = BIT(15);  /* This is response */
+
+	net_pkt_append_be16(pkt, dns_id);
+	net_pkt_append_be16(pkt, flags);   /* Flags and codes */
+	net_pkt_append_be16(pkt, 1);       /* Question count is set to 1, see
+					    * (RFC 4795 ch 2.1.1)
+					    */
+	net_pkt_append_be16(pkt, answers); /* Answer RR count */
+	net_pkt_append_be16(pkt, 0);       /* Authority RR count */
+	net_pkt_append_be16(pkt, 0);       /* Additional RR count */
+}
+
+static int add_question(struct net_pkt *pkt, enum dns_rr_type qtype,
+			struct net_buf *query)
+{
+	char *dot = query->data;
+	char *prev = NULL;
+
+	while ((dot = strchr(dot, '.'))) {
+		if (!prev) {
+			prev = dot++;
+			continue;
+		}
+
+		*prev = dot - prev - 1;
+		prev = dot++;
+	}
+
+	if (prev) {
+		*prev = strlen(prev) - 1;
+	}
+
+	if (!net_pkt_append_all(pkt, query->len + 1, query->data,
+				BUF_ALLOC_TIMEOUT)) {
+		return -ENOMEM;
+	}
+
+	net_pkt_append_be16(pkt, qtype);
+
+	net_pkt_append_be16(pkt, DNS_CLASS_IN);
+
+	return 0;
+}
+
+static int add_answer(struct net_pkt *pkt, enum dns_rr_type qtype,
+		      struct net_buf *query, u32_t ttl,
+		      u16_t addr_len, const u8_t *addr)
+{
+	add_question(pkt, qtype, query);
+
+	net_pkt_append_be32(pkt, ttl);
+	net_pkt_append_be16(pkt, addr_len);
+
+	if (!net_pkt_append_all(pkt, addr_len, addr,
+				BUF_ALLOC_TIMEOUT)) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static struct net_pkt *create_answer(struct net_context *ctx,
+				     sa_family_t family,
+				     enum dns_rr_type qtype,
+				     struct net_buf *query,
+				     u16_t dns_id,
+				     u16_t addr_len, const u8_t *addr)
+{
+	struct net_pkt *pkt;
+
+	pkt = net_pkt_get_tx(ctx, BUF_ALLOC_TIMEOUT);
+	if (!pkt) {
+		return NULL;
+	}
+
+	net_pkt_set_family(pkt, family);
+
+	setup_dns_hdr(pkt, 1, dns_id);
+
+	add_question(pkt, qtype, query);
+	add_answer(pkt, qtype, query, LLMNR_TTL, addr_len, addr);
+
+	return pkt;
+}
+
+#if defined(CONFIG_NET_IPV4)
+static const u8_t *get_ipv4_src(struct net_if *iface, struct in_addr *dst)
+{
+	const struct in_addr *addr;
+
+	addr = net_if_ipv4_select_src_addr(iface, dst);
+	if (!addr || net_is_ipv4_addr_unspecified(addr)) {
+		return NULL;
+	}
+
+	return (const u8_t *)addr;
+}
+#endif /* CONFIG_NET_IPV4 */
+
+#if defined(CONFIG_NET_IPV6)
+static const u8_t *get_ipv6_src(struct net_if *iface, struct in6_addr *dst)
+{
+	const struct in6_addr *addr;
+
+	addr = net_if_ipv6_select_src_addr(iface, dst);
+	if (!addr || net_is_ipv6_addr_unspecified(addr)) {
+		return NULL;
+	}
+
+	return (const u8_t *)addr;
+}
+#endif /* CONFIG_NET_IPV6 */
+
+static int send_response(struct net_context *ctx, struct net_pkt *pkt,
+			 struct net_buf *query, enum dns_rr_type qtype,
+			 u16_t dns_id)
+{
+	struct net_pkt *reply;
+	struct sockaddr dst;
+	socklen_t dst_len;
+	int ret;
+
+	if (net_pkt_family(pkt) == AF_INET) {
```

Comment:
looks like the code below could be factorized quite a bit actually, getting the addr according to query type mostly.

Reply:
Yes, this was the tricky part of the code. Looks a bit hairy indeed. Will investigate.